client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java

```diff
@@ -951,6 +1001,61 @@ public void testApiNamingConventions() throws Exception {
         assertThat("Some API are not supported but they should be: " + apiUnsupported, apiUnsupported.size(), equalTo(0));
     }
 
+    private static void doTestProductCompatibilityCheck(
+        boolean shouldBeAccepted, String version, boolean setProductHeader) throws Exception {
+
+        // An endpoint different from "/" that returns a boolean
+        GetSourceRequest apiRequest = new GetSourceRequest("foo", "bar");
+
+        StatusLine apiStatus = mock(StatusLine.class);
+        when(apiStatus.getStatusCode()).thenReturn(200);
+
+        Response apiResponse = mock(Response.class);
+        when(apiResponse.getStatusLine()).thenReturn(apiStatus);
+
+        RestClient restClient = mock(RestClient.class);
+
+        Build build = new Build(Build.Flavor.DEFAULT, Build.Type.UNKNOWN, "hash", "date", false, version);
+        mockGetRoot(restClient, build, setProductHeader);
+        when(restClient.performRequest(argThat(new RequestMatcher("HEAD", "/foo/_source/bar")))).thenReturn(apiResponse);
+
+        RestHighLevelClient highLevelClient =  new RestHighLevelClient(restClient, RestClient::close, Collections.emptyList());
+
+        if (shouldBeAccepted) {
+            assertTrue(highLevelClient.existsSource(apiRequest, RequestOptions.DEFAULT));
+        } else {
+            expectThrows(ElasticsearchException.class, () ->
+                highLevelClient.existsSource(apiRequest, RequestOptions.DEFAULT)
+            );
+        }
+    }
+
+    public void testProductCompatibilityCheck() throws Exception {
```

Comment:
I think we need a check for when the root request is exceptional, throwing either a 401 or 403, and also if it throws something other than those status requests.

Reply:
See d981a15 for 401 & 403. Other failures of the info request are propagated as failures for the original API request. If `/` returns something other than 200, 401 or 403 then we have a bigger problem and we won't let API requests through, both because verification could not happen and also because they will most likely fail too.